src/analyses/ai_history.cpp

```diff
@@ -27,3 +27,169 @@ xmlt ai_history_baset::output_xml(void) const
   xml.data = out.str();
   return xml;
 }
+
+ai_history_baset::step_returnt
+call_stack_historyt::step(locationt to, const trace_sett &others) const
+{
+  DATA_INVARIANT(current_stack != nullptr, "current_stack must exist");
+
+  cse_ptrt next_stack = nullptr;
+
+  // First construct what the new history would be.
+  // This requires special handling if this edge is a function call or return
+  if(current_stack->current_location->is_function_call())
+  {
+    locationt l_return = std::next(current_stack->current_location);
+
+    if(l_return->location_number == to->location_number)
+    {
+      // Is skipping the function call, only update the location
+      next_stack =
+        cse_ptrt(new call_stack_entryt(l_return, current_stack->caller));
+    }
+    else
+    {
+      // An interprocedural (call) edge; add to the current stack
+      next_stack = cse_ptrt(new call_stack_entryt(to, current_stack));
+    }
+  }
+  else if(current_stack->current_location->is_end_function())
+  {
+    if(current_stack->caller == nullptr)
+    {
+      // Trying to return but there was no caller?
+      // Refuse to do the transition outright
+      return std::make_pair(ai_history_baset::step_statust::BLOCKED, nullptr);
+    }
+    else
+    {
+      // The expected call return site...
+      locationt l_caller_return =
+        std::next(current_stack->caller->current_location);
+
+      if(l_caller_return->location_number == to->location_number)
+      {
+        // ... which is where we are going
+        next_stack =
+          cse_ptrt(new call_stack_entryt(to, current_stack->caller->caller));
+      }
+      else
+      {
+        // Not possible to return to somewhere other than the call site
```

Comment:
It's not an error.  The `ait` dispatch loop doesn't know which transitions are possible so it simply tries all of them (in this case, trying all transitions from *any* history that reaches the end of the callee function back to the call site) and relies on the history to block out any of these that the history can rule impossible.

This is kind of subtle and the first time this feature has been used; that's sort of why I wanted to write this patch to demonstrate this.

Reply:
Probably out of scope of this PR, but how would all this work with exceptions or something like setjmp/longjmp ?

Reply:
At the moment, it doesn't.  `ait` will through an exception if it finds try `THROW` or `CATCH` instructions and some domains will also fail in transform if they meet them.  So the following is hypothetical but...

(-2. Find a use case for analysing setjmp/longjmp. )
(-1. Get setjmp/longjmp working with the rest of the code. )
0. Actually get some agreement and documentation of the semantics of THROW/CATCH.
1. Inherit from `ait` and create an abstract interpreter that doesn't use recursion for interprocedural stuff.  This will need to track call-sites for functions.
2. Add tracking of catch locations and override `visit` to handle throw and catch.  Throw will then attempt to step edges to all possible catch locations.
3. Modify this, or another history so that it blocks steps to implausible exception handlers (i.e. not in the call stack) as given above.
4. Modify domains as needed so they can compute the edge.

Thinking about this ... actually the part from 1 to 4 is probably a single PR's worth.  I suspect the hard work is up to and including 0.